### PR TITLE
Fixed 'hidden' form type deprecation warning

### PR DESCRIPTION
**Pull request by verheyenkoen**
I am targeting this branch, because this is just a fix for a deprecation warning. Change is BC.

## Changelog

```markdown
### Fixed
- Fixed deprecation for Sf 3. support
```

## Subject

Fixed a Symfony 3 form type as string deprecation warning.

### CODE REVIEW

**Comment by verheyenkoen** [comment]
Hmmz, seems like there's a reason this wasn't fixed yet (PHP 5.3). Guess there's no other way than to live with the deprecation warning then?

File: Admin/Extension/LockExtension.php [hunk 1]
```diff
@@ -57,7 +58,7 @@ public function configureFormFields(FormMapper $form)
                 return;
             }
 
-            $form->add($fieldName, 'hidden', array(
+            $form->add($fieldName, HiddenType::class, array(
```
Comment (OskarStark) [inline Admin/Extension/LockExtension.php]: please do not use the `::class` constant and make it compatible with SF <2.8

like here for example: https://github.com/sonata-project/SonataPageBundle/pull/831/files

File: Admin/Extension/LockExtension.php [hunk 2]
```diff
@@ -57,10 +57,17 @@ public function configureFormFields(FormMapper $form)
                 return;
             }
 
-            $form->add($fieldName, 'hidden', array(
-                'mapped' => false,
-                'data' => $lockVersion,
-            ));
+            $form->add(
+                $fieldName,
+                // NEXT_MAJOR: remove these three lines and uncomment the one following
```
Comment (OskarStark) [inline Admin/Extension/LockExtension.php]: `NEXT_MAJOR: remove the check and add the FQCN`

**Comment by OskarStark** [comment]
Can you please check travis?

File: Admin/Extension/LockExtension.php [hunk 3]
```diff
@@ -57,10 +57,17 @@ public function configureFormFields(FormMapper $form)
...
 
-            $form->add($fieldName, 'hidden', array(
-                'mapped' => false,
-                'data' => $lockVersion,
-            ));
+            $form->add(
+                $fieldName,
+                // NEXT_MAJOR: remove the check and add the FQCN
+                method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix') ?
+                    'Symfony\Component\Form\Extension\Core\Type\HiddenType' :
```
Comment (Soullivaneuh) [inline Admin/Extension/LockExtension.php]: Please put `?` here and `:` below.

File: Tests/Admin/Extension/LockExtensionTest.php [hunk 4]
```diff
@@ -66,11 +66,18 @@ public function testConfigureFormFields()
     {
         $formMapper = $this->configureFormMapper();
         $form = $this->configureForm();
+
```
Comment (Soullivaneuh) [inline Tests/Admin/Extension/LockExtensionTest.php]: Please remove this new line. Not part of the PR.

File: Tests/Admin/Extension/LockExtensionTest.php [hunk 5]
```diff
@@ -66,11 +66,18 @@ public function testConfigureFormFields()
...
         $this->configureAdmin(null, null, $this->modelManager->reveal());
         $event = new FormEvent($form->reveal(), array());
 
         $this->modelManager->getLockVersion(array())->willReturn(1);
-        $form->add('_lock_version', 'hidden', array('mapped' => false, 'data' => 1))->shouldBeCalled();
+
+        // NEXT_MAJOR: remove the check and add the FQCN
+        $hiddenType = method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix') ?
```
Comment (Soullivaneuh) [inline Tests/Admin/Extension/LockExtensionTest.php]: Please don't use intermediate variables.

Comment (verheyenkoen) [inline Tests/Admin/Extension/LockExtensionTest.php]: Should I add the whole ternary operator as a method argument then?

File: Admin/Extension/LockExtension.php (hunk 3 above)
Comment (verheyenkoen) [inline Admin/Extension/LockExtension.php]: Don't understand. Please clarify.

File: Tests/Admin/Extension/LockExtensionTest.php (hunk 5 above)
Comment (Soullivaneuh) [inline Tests/Admin/Extension/LockExtensionTest.php]: Exactly. :wink:

Comment (verheyenkoen) [inline Tests/Admin/Extension/LockExtensionTest.php]: I'm wondering whether this `method_exists` approach is performing well. I find this particular statement 115 times in SonataAdminBundle alone (including tests that is but still):

```
method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')
```

Shouldn't this be extracted as a global variable or something (out of scope of this PR obviously) or can we assume `method_exists` is optimized for this?

Comment (greg0ire) [inline Tests/Admin/Extension/LockExtensionTest.php]: I think we should prove with a benchmark that this is inefficient before considering anything like that. IMO it's probably ok.

Comment (greg0ire) [inline Tests/Admin/Extension/LockExtensionTest.php]: See https://github.com/phpbench/phpbench if you want to try that nice tool

Comment (verheyenkoen) [inline Tests/Admin/Extension/LockExtensionTest.php]: @greg0ire Do you mean a benchmark on this approach as used in SonataAdminBundle or would a quick benchmark on `method_exists` be sufficient?

Comment (greg0ire) [inline Tests/Admin/Extension/LockExtensionTest.php]: A quick benchmark of 1000 calls to method_exists vs 1 call + 999 calls to a global variable should do.

Comment (verheyenkoen) [inline Tests/Admin/Extension/LockExtensionTest.php]: I tried this but the results seem to be within the margin of error. So no issue here I guess.

File: Admin/Extension/LockExtension.php (hunk 3 above)
Comment (jlamur) [inline Admin/Extension/LockExtension.php]: He asks you to do this:

```
method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')
     ? 'Symfony\Component\Form\Extension\Core\Type\HiddenType'
     : 'hidden',
```

**Comment by greg0ire** [comment]
Can you please get rid of the merge commit by rebasing on `upstream/3.x`?

**Comment by verheyenkoen** [comment]
@greg0ire I'm afraid I lack some git skills for that. I'd think that is not possible without rewriting git history (which I was told is not done).

**Comment by greg0ire** [comment]
>  I lack some git skills for that

Well learn them!

1. `git rebase upstream/3.x`;
2. `git push --force`.

> which I was told is not done

It's forbidden to do that on merged code. But no one uses the code you wrote here yet, you can rewrite it all you want.

**Comment by verheyenkoen** [comment]
@greg0ire I guess I f*cked up. A whole new block of expectations appears in the test file and I don't know where they come from.

**Comment by greg0ire** [comment]
What about the diff? Does it look good? Because said block might have been written by someone else.

**Comment by verheyenkoen** [comment]
@greg0ire It concerns file Tests/Admin/Extension/LockExtensionTest.php lines 72 until 94. I guess if someone else wrote them and they came with the rebase, they shouldn't appear as added in the diff?

**Comment by greg0ire** [comment]
No they should not indeed.

File: Tests/Admin/Extension/LockExtensionTest.php [hunk 6]
```diff
@@ -66,35 +66,19 @@ public function testConfigureFormFields()
     {
         $formMapper = $this->configureFormMapper();
         $form = $this->configureForm();
-        $event = new FormEvent($form, array());
 
-        $this->modelManager->expects($this->once())
-            ->method('getLockVersion')
-            ->will($this->returnValue(1));
+        $this->configureAdmin(null, null, $this->modelManager->reveal());
```
Comment (greg0ire) [inline Tests/Admin/Extension/LockExtensionTest.php]: @verheyenkoen this is where it comes from: a merge. Learn interactive rebase, get rid of this commit and you should be good.

**Comment by verheyenkoen** [comment]
Finally managed to get it right I think. Anything else I need to do or are we just waiting for @Soullivaneuh 's approval?

**Comment by greg0ire** [comment]
Thanks @verheyenkoen !
